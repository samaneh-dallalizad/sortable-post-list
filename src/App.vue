<template>
  <section class="main_area">
    <div class="container mx-auto">
      <div v-if="error === undefined && posts.length > 0" class="lg:flex">
        <!--PostLists-->
        <aside class="lg:w-1/2 px-10 pt-20">
          <h2 class="text-white text-2xl pb-4 ">Sortable Post List </h2>
          <div id="post_list">
            <ul class="flex flex-col space-y-4">
              <PostsList
                  :posts="posts"
                  @changeOrder="changeOrder"
              />
            </ul>
          </div>
        </aside>
        <!--ActionLists-->
        <aside class="lg:w-1/2 pt-10 px-20">
          <div class="action_list">
            <div class="bg-white p-5 rounded-t-md"><h3 class="text-[#626262]">List of actions commited</h3></div>
            <div class="rounded-md shadow-lg p-4">
              <ul>
                <ActionsList
                    :actions="actionHistory"
                    @timeTravel="timeTravel"
                />
              </ul>
            </div>
          </div>
        </aside>
      </div>
      <div v-else class="flex justify-center">
        <div class="error-msg">
          <div role="alert">
            <p class="font-bold">Informational message</p>
            <p class="text-sm">{{ this.error }}</p>
          </div>
        </div>
      </div>
    </div>
  </section>

</template>

<script>
import PostsList from './components/PostsList.vue'
import ActionsList from './components/ActionsList.vue'

export default {
  name: 'App',
  components: {
    PostsList,
    ActionsList
  },
  data() {
    return {
      dataUrl: "https://jsonplaceholder.typicode.com/posts",
      posts: [],
      actionHistory: [],
      error: undefined
    }
  },
  mounted() {
    this.getData();
  },
  methods: {
    getData: async function () {
      let response = await fetch(this.dataUrl)
      if (response.ok === false) {
        this.error = "Cannot fetch posts"
      } else {
        let data = await response.json()
        this.posts = data.slice(0, 5)
      }
    },
    changeOrder(direction, index) {
      let tmp = this.posts[index]
      switch (direction) {
        case 'down':
          this.posts[index] = this.posts[index + 1]
          this.posts[index + 1] = tmp
          this.actionHistory.unshift({from: index, to: index + 1, postId: tmp.id})
          break
        case 'up':
          this.posts[index] = this.posts[index - 1]
          this.posts[index - 1] = tmp
          this.actionHistory.unshift({from: index, to: index - 1, postId: tmp.id})
          break
      }

    },
    timeTravel(index) {
      for (let i = 0; i <= index; i++) {
        let tmp = this.posts[this.actionHistory[i].from]
        this.posts[this.actionHistory[i].from] = this.posts[this.actionHistory[i].to]
        this.posts[this.actionHistory[i].to] = tmp
      }
      this.actionHistory = this.actionHistory.slice(index + 1);
    },

  }
}
</script>
<style scoped>
.main_area {
  position: relative;
}

.main_area:before {
  position: absolute;
  top: 0;
  left: 0;
  content: "";
  width: 0;
  height: 0;
  border-top: 230px solid #6357b1;
  border-right: 100vw solid transparent;
  z-index: -1;
}

.action_list {
  @apply rounded-md shadow-lg bg-[#f4f4f4];
}

.action_list ul {
  @apply grid grid-cols-1 divide-y text-sm text-center mx-auto shadow-lg;
}

.error-msg {
  @apply bg-red-100 border-t border-b border-red-500 text-red-700 px-4 py-3 my-8 md:w-6/12 sm:w-full;
}
</style>


