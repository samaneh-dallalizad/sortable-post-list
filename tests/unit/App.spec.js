import {shallowMount} from '@vue/test-utils'
import App from '@/App.vue'

describe('App.vue', () => {
    it('will show errors', () => {
        const wrapper = shallowMount(App, {
            data() {
                return {
                    posts: [],
                    actionHistory: [],
                    error: "Cannot fetch posts"
                }
            },
            mounted() {
            }
        })
        expect(wrapper.element).toMatchSnapshot()
    })

    it('will show skeleton', () => {
        const wrapper = shallowMount(App, {
            data() {
                return {
                    posts: [{id: 1}],
                    actionHistory: []
                }
            },
            mounted() {
            }
        })
        expect(wrapper.element).toMatchSnapshot()
    })

    it('changes order', () => {
        const wrapper = shallowMount(App, {
            data() {
                return {
                    posts: [{id: 1}, {id: 2}, {id: 3}],
                    actionHistory: []
                }
            },
            mounted() {
            }
        })
        wrapper.vm.changeOrder('down', 0)
        wrapper.vm.changeOrder('up', 2)
        expect(wrapper.vm.posts).toEqual([{id: 2}, {id: 3}, {id: 1}])
    })

    it('time travel', () => {
        const wrapper = shallowMount(App, {
            data() {
                return {
                    posts: [{id: 3}, {id: 2}, {id: 1}],
                    actionHistory: [{from: 0, to: 1, postId: 1}, {from: 1, to: 2, postId: 1}, {
                        from: 1,
                        to: 0,
                        postId: 3
                    }]
                }
            },
            mounted() {
            }
        })

        wrapper.vm.timeTravel(wrapper.vm.actionHistory.length - 1)

        expect(wrapper.vm.actionHistory).toEqual([])
        expect(wrapper.vm.posts).toEqual([{id: 1}, {id: 2}, {id: 3}])
    })

})
