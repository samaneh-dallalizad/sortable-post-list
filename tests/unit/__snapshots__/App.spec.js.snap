// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.vue will show errors 1`] = `
<section
  class="main_area"
>
  <div
    class="container mx-auto"
  >
    <div
      class="flex justify-center"
    >
      <div
        class="error-msg"
      >
        <div
          role="alert"
        >
          <p
            class="font-bold"
          >
            Informational message
          </p>
          <p
            class="text-sm"
          >
            Cannot fetch posts
          </p>
        </div>
      </div>
    </div>
  </div>
</section>
`;

exports[`App.vue will show skeleton 1`] = `
<section
  class="main_area"
>
  <div
    class="container mx-auto"
  >
    <div
      class="lg:flex"
    >
      <!--PostLists-->
      <aside
        class="lg:w-1/2 px-10 pt-20"
      >
        <h2
          class="text-white text-2xl pb-4"
        >
          Sortable Post List 
        </h2>
        <div
          id="post_list"
        >
          <ul
            class="flex flex-col space-y-4"
          >
            <posts-list-stub
              posts="[object Object]"
            />
          </ul>
        </div>
      </aside>
      <!--ActionLists-->
      <aside
        class="lg:w-1/2 pt-10 px-20"
      >
        <div
          class="action_list"
        >
          <div
            class="bg-white p-5 rounded-t-md"
          >
            <h3
              class="text-[#626262]"
            >
              List of actions commited
            </h3>
          </div>
          <div
            class="rounded-md shadow-lg p-4"
          >
            <ul>
              <actions-list-stub
                actions=""
              />
            </ul>
          </div>
        </div>
      </aside>
    </div>
  </div>
</section>
`;
